components/scream/src/physics/p3/p3_functions_f90.cpp

```diff
@@ -460,20 +451,88 @@ void calc_first_order_upwind_step_f_impl(
   });
 
   // Sync back to host
-  for (Int i = 0; i < N; ++i) {
-    Kokkos::deep_copy(fluxes_h[i], fluxes_d[i]);
-    Kokkos::deep_copy(vs_h[i],     vs_d[i]);
-    Kokkos::deep_copy(qnx_h[i],    qnx_d[i]);
-
-    for (Int k = 0; k < npack; ++k) {
-      const Int scalar_offset = k*Spack::n;
-      for (Int s = 0; s < Spack::n && scalar_offset+s < nk; ++s) {
-        fluxes[i][scalar_offset + s] = fluxes_h[i](k)[s];
-        vs    [i][scalar_offset + s] = vs_h    [i](k)[s];
-        qnx   [i][scalar_offset + s] = qnx_h   [i](k)[s];
-      }
+  pack::device_to_host<N, Real, Spack, DefaultDevice>(ptr_to_arr<N>(fluxes), nk, fluxes_d);
+  pack::device_to_host<N, Real, Spack, DefaultDevice>(ptr_to_arr<N>(qnx), nk, qnx_d);
+}
+
+template <int N>
+void generalized_sedimentation_f_impl(
+  Int kts, Int kte, Int kdir, Int k_qxtop, Int* k_qxbot, Int kbot, Real Co_max,
+  Real* dt_left, Real* prt_accum, Real* inv_dzq, Real* inv_rho, Real* rho,
+  Real** vs, Real** fluxes, Real** qnx)
+{
+  using P3F  = Functions<Real, DefaultDevice>;
+
+  using Spack = typename P3F::Spack;
+  using Singlep = typename pack::Pack<Real, 1>;
+  using view_1d = typename P3F::view_1d<Spack>;
+  using view_1ds = typename P3F::view_1d<Singlep>;
+  using KT = typename P3F::KT;
+  using ExeSpace = typename KT::ExeSpace;
+  using MemberType = typename P3F::MemberType;
+  using view_1d_ptr_array = typename P3F::view_1d_ptr_array<Spack, N>;
+  using uview_1d = typename P3F::uview_1d<Spack>;
+
+  scream_require_msg(kts == 1, "kts must be 1, got " << kts);
+
+  // Adjust for 0-based indexing
+  kts -= 1;
+  kte -= 1;
+  kbot -= 1;
+  k_qxtop -= 1;
+  *k_qxbot -= 1;
+
+  const Int nk = (kte - kts) + 1;
+
+  // Setup views
+  Kokkos::Array<view_1d, 3> temp_d;
+  Kokkos::Array<view_1d, N> fluxes_d, vs_d, qnx_d;
+  Kokkos::Array<view_1ds, 1> scalar_temp;
+  std::vector<Real> scalars = {*prt_accum, *dt_left, static_cast<Real>(*k_qxbot)};
+
+  pack::host_to_device<3, Real, Spack, DefaultDevice>({rho, inv_rho, inv_dzq}, nk, temp_d);
+  pack::host_to_device<1, Real, Singlep, DefaultDevice>({scalars.data()}, scalars.size(), scalar_temp);
+
+  view_1d rho_d(temp_d[0]), inv_rho_d(temp_d[1]), inv_dzq_d(temp_d[2]);
+  view_1ds scalars_d(scalar_temp[0]);
```

Comment:
Why is `scalars_d`'s value type a `Spack` rather than a `Real`?

Reply:
So I can use pack::host_to_device, which expects a pack type, so I used Pack<1>

Reply:
I see, and since this is testing code, such shortcuts are permitted.